### PR TITLE
Deduplicate single/double precision LAPACK

### PR DESCRIPTION
**Pull request by pazner**
Deduplicate calls to BLAS/LAPACK using the prefix macro `MFEM_LAPACK_PREFIX` (and `MFEM_LAPACK_COMPLEX` for complex numbers).

This simplifies the code significantly (much less preprocessor branching) and eliminates the risk of modifying only one branch of the `#ifdef MFEM_USE_DOUBLE`.

Also, adjust some hard-coded tolerances so that ex38 works in single-precision now.

 | PR | Author | Editor | Reviewers | Assignment | Approval | Merge|
 | --- | --- | --- | --- | --- | --- | --- |
| [#4337](https://github.com/mfem/mfem/pull/4337) | @pazner | @tzanio | @dylan-copeland + @psocratis | 6/2/24 | 6/6/24 | 6/11/24 | |

<details>
<summary>PR Checklist</summary>
    
- [ ] Code builds.
- [ ] Code passes `make style`.
- [ ] Update `CHANGELOG`:
    - [ ] Is this a new feature users need to be aware of? New or updated example or miniapp?
    - [ ] Does it make sense to create a new section in the `CHANGELOG` to group with other related features?
- [ ] Update `INSTALL`:
    - [ ] Had a new optional library been added? If so, what range of versions of this library are required? (*Make sure the external library is compatible with our BSD license, e.g. it is not licensed under GPL!*)
    - [ ] Have the version ranges for any required or optional libraries changed?
    - [ ] Does `make` or `cmake` have a new target?
    - [ ] Did the requirements or the installation process change? *(rare)*
- [ ] Update continuous integration server configurations if necessary (e.g. with new version requirements for each of MFEM's dependencies)
    - [ ] `.github`
    - [ ] `.appveyor.yml`
- [ ] Update `.gitignore`:
    - [ ] Check if `make distclean; git status` shows any files that were generated from the source by the project (not an IDE) but we don't want to track in the repository.
    - [ ] Add new patterns (just for the new files above) and re-run the above test.
- [ ] New examples:
    - [ ] All sample runs at the top of the example source file work.
    - [ ] Update `examples/makefile`:
      - [ ] Add the example code to the appropriate `SEQ_EXAMPLES` and `PAR_EXAMPLES` variables.
      - [ ] Add any files generated by it to the `clean` target.
      - [ ] Add the example binary and any files generated by it to the top-level `.gitignore` file.
    - [ ] Update `examples/CMakeLists.txt`:
      - [ ] Add the example code to the `ALL_EXE_SRCS` variable.
      - [ ] Make sure `THIS_TEST_OPTIONS` is set correctly for the new example.
   - [ ] List the new example in `doc/CodeDocumentation.dox`.
   - [ ] If new examples directory (e.g.`examples/pumi`), list it in `doc/CodeDocumentation.conf.in`
   - [ ] Companion pull request for documentation in [mfem/web](https://github.com/mfem/web) repo:
      - [ ] Update or add example-specific documentation, see e.g. the `src/examples.md`.
      - [ ] Add the description, labels and screenshots in `src/examples.md` and `src/img`.
      - [ ] In `examples.md`, list the example under the appropriate categories, add new categories if necessary.
      - [ ] Add a short description of the example in the "Extensive Examples" section of `features.md`.
- [ ] New miniapps:
   - [ ] All sample runs at the top of the miniapp source file work.
   - [ ] Update top-level `makefile` and `makefile` in corresponding miniapp directory.
   - [ ] Add the miniapp binary and any files generated by it to the top-level `.gitignore` file.
   - [ ] Update CMake build system:
      - [ ] Update the `CMakeLists.txt` file in the `miniapps` directory, if the new miniapp is in a new directory.
      - [ ] Add/update the `CMakeLists.txt` file in the new miniapp directory.
      - [ ] Consider adding a new test for the new miniapp.
   - [ ] List the new miniapp in `doc/CodeDocumentation.dox`
   - [ ] If new miniapps directory (e.g.`miniapps/nurbs`), add it to `MINIAPP_SUBDIRS` in the `makefile`.
   - [ ] If new miniapps directory (e.g.`miniapps/nurbs`), list it in `doc/CodeDocumentation.conf.in`
   - [ ] Companion pull request for documentation in [mfem/web](https://github.com/mfem/web) repo:
     - [ ] Update or add miniapp-specific documentation, see e.g. the `src/meshing.md` and `src/electromagnetics.md` files.
     - [ ] Add the description, labels and screenshots in `src/examples.md` and `src/img`.
     - [ ] The miniapps go at the end of the page, and are usually listed only under a specific "Application (PDE)" category.
     - [ ] Add a short description of the miniapp in the "Extensive Examples" section of `features.md`.
- [ ] New capability:
   - [ ] All new public, protected, and private classes, methods, data members, and functions have full Doxygen-style documentation in source comments. Documentation should include descriptions of member data, function arguments and return values, template parameters, and prerequisites for calling new functions.
   - [ ] Pointer arguments and return values must specify whether ownership is being transferred or lent with the call.
   - [ ] Any new functions should include descriptions of their intended use e.g. for internal use only, user-facing, etc., along with references to example code whenever possible/appropriate.
   - [ ] Consider adding new sample runs in existing examples to highlight the new capability.
   - [ ] Consider saving cool simulation pictures with the new capability in the Confluence gallery (LLNL only) or submitting them, via pull request, to the gallery section of the `mfem/web` repo.
   - [ ] If this is a major new feature, consider mentioning it in the short summary inside `README` *(rare)*.
   - [ ] List major new classes in `doc/CodeDocumentation.dox` *(rare)*.
- [ ] Update this checklist, if the new pull request affects it.
- [ ] Run `make unittest` to make sure all unit tests pass.
- [ ] Run the tests in `tests/scripts`.
- [ ] (LLNL only) After merging:
   - [ ] Update internal tests to include the new features.
</details>

### CODE REVIEW

**Comment by tzanio** [comment]
This PR is now [under review](https://github.com/mfem/mfem/blob/master/CONTRIBUTING.md#pull-requests) (see the table in the PR description). To help with the review process, please **do not force push** to the branch.

File: linalg/solvers.cpp [hunk 1]
```diff
@@ -3938,25 +3907,17 @@ void NNLSSolver::Solve(const Vector& rhs_lb, const Vector& rhs_ub,
...
-            sormqr_(&lside, &trans, &m, &n_update, &i_qr_start,
-#elif defined MFEM_USE_DOUBLE
-            dormqr_(&lside, &trans, &m, &n_update, &i_qr_start,
-#endif
-                    mat_qr_data.GetData(), &m, tau.GetData(),
-                    mat_qr_data.GetData() + (i_qr_start * m), &m,
-                    work.data(), &lwork, &info);
+            MFEM_LAPACK_PREFIX(ormqr_)(&lside, &trans, &m, &n_update, &i_qr_start,
+                                       mat_qr_data.GetData(), &m, tau.GetData(),
+                                       mat_qr_data.GetData() + (i_qr_start * m), &m,
```
Comment (dylan-copeland) [inline linalg/solvers.cpp:3912]: Slightly exceeding 80 columns here and below.

File: linalg/densemat.cpp [hunk 2]
```diff
@@ -1066,31 +953,19 @@ void dsyevr_Eigensystem(DenseMatrix &a, Vector &ev, DenseMatrix *evect)
...
-#elif defined MFEM_USE_DOUBLE
-   dsyevr_( &JOBZ, &RANGE, &UPLO, &N, A, &LDA, &VL, &VU, &IL, &IU,
-#else
-   MFEM_ABORT("Floating point type undefined");
-#endif
-            &ABSTOL, &M, W, Z, &LDZ, ISUPPZ, &QWORK, &LWORK,
-            &QIWORK, &LIWORK, &INFO );
+   MFEM_LAPACK_PREFIX(syevr_)(&JOBZ, &RANGE, &UPLO, &N, A, &LDA, &VL, &VU, &IL,
+                              &IU, &ABSTOL, &M, W, Z, &LDZ, ISUPPZ, &QWORK,
+                              &LWORK, &QIWORK, &LIWORK, &INFO );
```
Comment (dylan-copeland) [inline linalg/densemat.cpp:958]: Extra space after `INFO` can be removed, here and below.

File: linalg/densemat.cpp [hunk 3]
```diff
@@ -4010,7 +3789,8 @@ void CholeskyFactors::LSolve(int m, int n, real_t * X) const
    int info = 0;
 
 #ifdef MFEM_USE_SINGLE
-   strtrs_(&uplo, &trans, &diag, &m, &n, data, &m, X, &m, &info);
```
Comment (dylan-copeland) [inline linalg/densemat.cpp:4013]: `#ifdef MFEM_USE_SINGLE` should be removed (looks you only missed one!).

**Review by psocratis — APPROVED** [review]
After testing with `make test` with single precision and with MFEM_USE_LAPACK there were 2 issues not related to this PR's changes but still related to single precision. The fix is simple and so I pushed the small changes in this branch. 

Very nice clean up for the lapack interface!

File: linalg/solvers.cpp (hunk 1 above)
Comment (pazner) [reply linalg/solvers.cpp]: Fixed in e4b8584a16616f2b50422b85c5ba0fa30cd68c46

File: linalg/densemat.cpp (hunk 2 above)
Comment (pazner) [reply linalg/densemat.cpp]: Fixed in e4b8584a16616f2b50422b85c5ba0fa30cd68c46

File: linalg/densemat.cpp (hunk 3 above)
Comment (pazner) [reply linalg/densemat.cpp:4013]: Fixed in 262fa6173d8307be5e25bc2f945ef43dfd368174

File: linalg/densemat.cpp [hunk 4]
```diff
@@ -4010,7 +3789,8 @@ void CholeskyFactors::LSolve(int m, int n, real_t * X) const
...
+   MFEM_LAPACK_PREFIX(trtrs_)(&uplo, &trans, &diag, &m, &n, data, &m, X, &m,
+                              &info);
 #elif defined MFEM_USE_DOUBLE
```
[the remark this answers is not in the document]
Comment (pazner) [reply linalg/densemat.cpp]: Fixed in 262fa6173d8307be5e25bc2f945ef43dfd368174

**Review by dylan-copeland — APPROVED** [review]

**Comment by tzanio** [comment]
Merged in `next` for testing...

**Review by tzanio — APPROVED** [review]
Thanks for adding this @pazner! 🙏 

@psocratis and @dylan-copeland -- thanks for the reviews! 👍

**Pull request merged**